azure-mgmt-resources/src/main/java/com/azure/management/BearerTokenAuthenticationPolicy.java

```diff
@@ -0,0 +1,101 @@
+        Objects.requireNonNull(credential);
+        this.credential = credential;
+        this.scopes = scopes;
+        this.tokenCache = new HashMap<>();
+
+        if (credential instanceof AzureTokenCredential) {
+            this.environment = ((AzureTokenCredential) credential).getEnvironment();
+        } else {
+            this.environment = AzureEnvironment.AZURE;
+        }
+    }
+
+    private String getDefaultScopeFromRequest(HttpRequest request) {
+        String host = request.getUrl().getHost();
+        String resource = this.environment.getManagementEndpoint();
+        for (Map.Entry<String, String> endpoint : this.environment.endpoints().entrySet()) {
+            if (host.contains(endpoint.getValue())) {
+                if (endpoint.getKey().equals(AzureEnvironment.Endpoint.KEYVAULT.identifier())) {
+                    resource = String.format("https://%s/", endpoint.getValue().replaceAll("^\\.*", ""));
+                    break;
+                } else if (endpoint.getKey().equals(AzureEnvironment.Endpoint.GRAPH.identifier())) {
+                    resource = this.environment.getGraphEndpoint();
+                    break;
+                } else if (endpoint.getKey().equals(AzureEnvironment.Endpoint.LOG_ANALYTICS.identifier())) {
+                    resource = this.environment.getLogAnalyticsEndpoint();
+                    break;
+                } else if (endpoint.getKey().equals(AzureEnvironment.Endpoint.APPLICATION_INSIGHTS.identifier())) {
+                    resource = this.environment.getApplicationInsightsEndpoint();
+                    break;
+                } else if (endpoint.getKey().equals(AzureEnvironment.Endpoint.DATA_LAKE_STORE.identifier())
+                        || endpoint.getKey().equals(AzureEnvironment.Endpoint.DATA_LAKE_ANALYTICS.identifier())) {
+                    resource = this.environment.getDataLakeEndpointResourceId();
+                    break;
+                }
+            }
+        }
+        return resource + "/.default";
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        if ("http".equals(context.getHttpRequest().getUrl().getProtocol())) {
+            return Mono.error(new RuntimeException("token credentials require a URL using the HTTPS protocol scheme"));
+        }
+
+        String[] scopes;
+        if (this.scopes == null || this.scopes.length == 0) {
+            scopes = new String[] {getDefaultScopeFromRequest(context.getHttpRequest())};
+        } else {
+            scopes = this.scopes;
+        }
+        assert scopes.length > 0;
+
+        Mono<AccessToken> tokenResult;
+        AccessToken token = tokenCache.get(scopes[0]);
+        if (token == null || token.isExpired()) {
+            tokenResult = this.credential.getToken(new TokenRequestContext().addScopes(scopes)).doOnNext(accessToken -> this.tokenCache.put(scopes[0], accessToken));
```

Comment:
Not sure about `doOnNext`, it could still follow by an Error instead of Complete. Maybe `doOnSuccess`?

Reply:
I'll test about it, since we don't catch exception, if the credential get token with error it will throw it and the pipeline will be terminated.

Reply:
The point here is that I do not want to write to the cache unless it is a success (+ token not null).

Reply:
https://stackoverflow.com/questions/55976963/is-there-a-difference-between-doonsuccess-vs-doonnext-for-a-mono As SO said, it seems doOnNext is more likely to be used which would skip when data is null.